community/neo4j/src/test/java/org/neo4j/locking/QueryExecutionLocks.java

```diff
@@ -0,0 +1,964 @@
+import org.neo4j.storageengine.api.NodeItem;
+import org.neo4j.storageengine.api.PropertyItem;
+import org.neo4j.storageengine.api.RelationshipItem;
+import org.neo4j.storageengine.api.Token;
+import org.neo4j.storageengine.api.lock.ResourceType;
+import org.neo4j.storageengine.api.schema.PopulationProgress;
+import org.neo4j.test.rule.EmbeddedDatabaseRule;
+import org.neo4j.values.storable.Value;
+
+import static java.util.Arrays.asList;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+
+public class QueryExecutionLocks
+{
+
+    @Rule
+    public EmbeddedDatabaseRule databaseRule = new EmbeddedDatabaseRule();
+
+    @Test
+    public void noLocksTakenForQueryWithoutAnyIndexesUsage() throws Exception
+    {
+        String query = "MATCH (n) return count(n)";
+        List<LockOperationRecord> lockOperationRecords = traceQueryLocks( query );
+        assertThat( "Observed list of lock operations is: " + lockOperationRecords,
+                lockOperationRecords, is( empty() ) );
+    }
+
+    @Test
+    public void takeLabelLockForQueryWithIndexUsages() throws Exception
+    {
+        String labelName = "Human";
+        Label human = Label.label( labelName );
+        String propertyKey = "name";
+        createIndex( human, propertyKey );
+
+        try ( Transaction transaction = databaseRule.beginTx() )
+        {
+            Node node = databaseRule.createNode( human );
+            node.setProperty( propertyKey, RandomStringUtils.randomAscii( 10 ) );
+            transaction.success();
+        }
+
+        String query = "MATCH (n:" + labelName + ") where n." + propertyKey + " = \"Fry\" RETURN n ";
+
+        List<LockOperationRecord> lockOperationRecords = traceQueryLocks( query );
+        assertThat( "Observed list of lock operations is: " + lockOperationRecords,
+                lockOperationRecords, hasSize( 1 ) );
+
+        LockOperationRecord operationRecord = lockOperationRecords.get( 0 );
+        assertTrue( operationRecord.acquisition );
+        assertFalse( operationRecord.exclusive );
+        assertEquals( ResourceTypes.LABEL, operationRecord.resourceType );
+    }
+
+    @Test
+    public void reTakeLabelLockForQueryWithIndexUsagesWhenSchemaStateWasUpdatedDuringLockOperations() throws Exception
+    {
+        String labelName = "Robot";
+        Label human = Label.label( labelName );
+        String propertyKey = "name";
+        createIndex( human, propertyKey );
+
+        try ( Transaction transaction = databaseRule.beginTx() )
+        {
+            Node node = databaseRule.createNode( human );
+            node.setProperty( propertyKey, RandomStringUtils.randomAscii( 10 ) );
+            transaction.success();
+        }
```

Comment:
Small nit, the variable name says Bender is human (which did happen in one episode)